src/java.base/share/classes/jdk/internal/foreign/MemorySessionImpl.java

```diff
@@ -51,42 +52,76 @@
  * access is possible when a session is being closed (see {@link jdk.internal.misc.ScopedMemoryAccess}).
  */
-public abstract sealed class MemorySessionImpl
-        implements Scope
-        permits ConfinedSession, GlobalSession, SharedSession {
-    static final int OPEN = 0;
-    static final int CLOSED = -1;
+public final class MemorySessionImpl implements Scope {
+    private static final int NONCLOSEABLE = 1;
+    private static final int OPEN = 0;
+    private static final int CLOSED = -1;
 
-    static final VarHandle STATE;
-    static final int MAX_FORKS = Integer.MAX_VALUE;
-
-    static final ScopedMemoryAccess.ScopedAccessError ALREADY_CLOSED = new ScopedMemoryAccess.ScopedAccessError(MemorySessionImpl::alreadyClosed);
-    static final ScopedMemoryAccess.ScopedAccessError WRONG_THREAD = new ScopedMemoryAccess.ScopedAccessError(MemorySessionImpl::wrongThread);
     // This is the session of all zero-length memory segments
-    public static final MemorySessionImpl GLOBAL_SESSION = new GlobalSession();
+    public static final MemorySessionImpl GLOBAL_SESSION;
+    static {
+        GLOBAL_SESSION = new MemorySessionImpl(null, null, null);
+        GLOBAL_SESSION.state = NONCLOSEABLE;
+    }
 
-    final ResourceList resourceList;
-    final Thread owner;
-    int state = OPEN;
+    private static final VarHandle STATE;
+    private static final VarHandle ACQUIRE_COUNT;
+    private static final VarHandle ASYNC_RELEASE_COUNT;
+    private static final int MAX_FORKS = Integer.MAX_VALUE;
+
+    private static final ScopedMemoryAccess SCOPED_MEMORY_ACCESS = ScopedMemoryAccess.getScopedMemoryAccess();
+    private static final ScopedMemoryAccess.ScopedAccessError ALREADY_CLOSED = new ScopedMemoryAccess.ScopedAccessError(MemorySessionImpl::alreadyClosed);
+    private static final ScopedMemoryAccess.ScopedAccessError WRONG_THREAD = new ScopedMemoryAccess.ScopedAccessError(MemorySessionImpl::wrongThread);
+
+    private final Thread owner;
+    private final Object ref;
+    private final ResourceList resourceList;
+
+    // There are 4 important operations on a MemorySession: checkValidStateRaw,
+    // justClose, acquire0, release0.
+    //
+    // There are 3 kinds of MemorySession:
+    //
+    // Confined session - owner is not null:
+    // - Only the owner can access state and acquireCount.
+    // - Only the owner can invoke checkValidStateRaw, justClose, acquire0.
+    // - Any thread can invoke release0.
+    //
+    // Shared session - owner is null, state is not NONCLOSEABLE:
+    // - Any thread can access any field, as well as invoke any method.
+    // - justClose, acquire0, and release0 synchronize with themselves and with
+    //   each other using acquireCount.
+    //
+    // Implicit session - owner is null, state is NONCLOSEABLE:
+    // - No thread can invoke justClose.
+    // - All fields are constant.
+    // - Any thread can invoke checkValidStateRaw, acquire0 and release0.
+    @Stable
```

Comment:
Question: are you using Stable here because the state for shared/confined starts from 0 (default) state, and then has only ONE transition to closed? (which is compatible with the semantics of `Stable`)? If so, a comment is needed here, as this is subtle!

Reply:
Yes I am, added some more explanation.